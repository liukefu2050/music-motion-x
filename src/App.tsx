/* eslint-disable react/require-default-props */
/* eslint-disable react/forbid-prop-types */

import React, { Suspense } from "react"
import { StaticRouter, StaticRouterProps } from "react-router-dom/server"
import {
  QueryClient,
  QueryClientProvider,
  useQueryErrorResetBoundary,
} from "react-query"
import { BrowserRouter } from "react-router-dom"
import { ErrorBoundary } from "react-error-boundary"
import { Provider } from "react-redux"
import { RouteMatch, Routes, Route } from "react-router"
import ErrorFound from "@/components/ErrorPage"

import routes from "./routes"
import NotFound from "./pages/NotFound"
import PlayBar from "@/components/PlayBar"
import AppUpdateAvailable from "@/components/AppUpdateAvailable"

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      suspense: true,
    },
  },
})

const App = ({
  store,
  isServer,
  location,
  preloadedState,
}: {
  store: any
  isServer: boolean
  matchedRoutes?: RouteMatch[]
  location?: StaticRouterProps["location"]
  preloadedState: { [x: string]: any }
}) => {
  const { reset } = useQueryErrorResetBoundary()
  const content = (
    <>
      {/* <PageBack isBlack={false} /> */}
      <Suspense>
        <PlayBar />
      </Suspense>

      <Routes>
        {routes.map((r) => (
          <Route element={r.element} path={r.path} key={r.path}>
            {r.children &&
              r.children.length > 0 &&
              r.children.map((child) => (
                <Route
                  element={child.element}
                  path={child.path}
                  key={child.path}
                />
              ))}
          </Route>
        ))}
        <Route path="*" element={<NotFound />} />
      </Routes>
    </>
  )

  const IsomophicRouter = isServer ? (
    <StaticRouter location={location || ""}>{content}</StaticRouter>
  ) : (
    <BrowserRouter>
      <>
        <AppUpdateAvailable />
        {content}
      </>
    </BrowserRouter>
  )

  return (
    <ErrorBoundary
      onReset={reset}
      fallbackRender={({ resetErrorBoundary, error }) => (
        <ErrorFound resetErrorBoundary={resetErrorBoundary} error={error} />
      )}
    >
      <Provider store={store} serverState={preloadedState || {}}>
        <QueryClientProvider client={queryClient}>
          {IsomophicRouter}
        </QueryClientProvider>
      </Provider>
    </ErrorBoundary>
  )
}

export default App
