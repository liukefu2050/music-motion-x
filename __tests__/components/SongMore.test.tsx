/* eslint-disable testing-library/no-unnecessary-act */
import nock from "nock"
import { fireEvent, screen, act, waitFor } from "@testing-library/react"
import SongMore from "../../src/components/SongMore"
import App from "../test-utils/StoreProvider"
import renderWithModalRoot from "../test-utils/renderWithModalRoot"
// const sleep = () => new Promise((resolve) => setTimeout(resolve, 1000))
const IntersectionObserver = window.IntersectionObserver
const songId = 3

describe("SongMore", () => {
  beforeAll(() => {
    nock.disableNetConnect()
  })

  afterAll(() => {
    nock.cleanAll()
    nock.enableNetConnect()
    window.IntersectionObserver = IntersectionObserver
  })

  beforeEach(() => {
    // @ts-ignore
    window.IntersectionObserver = function (fn: any, opts: any) {
      return {
        observe: () => {
          act(() => {
            fn([{ intersectionRatio: 1 }], {
              unobserve: (ele: HTMLElement) => {},
            })
          })
        },
        unobserve: (ele: HTMLElement) => {},
      }
    }

    nock("https://81.69.200.140")
      .get(`/api/check/music?id=${songId}`)
      .reply(200, {
        success: true,
      })
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.useRealTimers()
  })

  it("render well when in viewpoint", async () => {
    const { asFragment } = renderWithModalRoot(
      <App>
        <SongMore
          albumName="abc"
          artistName="cdf"
          artistId={1}
          songName="ghj"
          albumId={2}
          songId={songId}
          imgUrl={"fake-url"}
        />
      </App>,
    )

    expect(asFragment()).toMatchSnapshot()

    act(() => {
      fireEvent.click(screen.getByTestId("song-more"))
      jest.runOnlyPendingTimers()
    })
    await waitFor(() => {
      expect(screen.getByText("下一首播放")).toBeInTheDocument()
    })
    expect(asFragment()).toMatchSnapshot()
    fireEvent.click(screen.getByText("下一首播放"))

    act(() => {
      fireEvent.click(screen.getByTestId("modal-mask"))
    })
  })

  it("could work well with intersection-observer polyfill", () => {
    // @ts-ignore
    window.IntersectionObserver = null

    const songId = 3
    const outClick = jest.fn()
    renderWithModalRoot(
      <App>
        <div onClick={outClick}>
          <SongMore
            albumName="abc"
            artistName="cdf"
            artistId={1}
            songName="ghj"
            albumId={2}
            songId={songId}
            imgUrl={"fake-url"}
          />
        </div>
      </App>,
    )
  })

  it("not trigger fetch when not in viewpoint", async () => {
    // @ts-ignore
    window.IntersectionObserver = function (fn: any, opts: any) {
      return {
        observe: () => {
          act(() => {
            fn([{ intersectionRatio: -0.6 }], {
              unobserve: (ele: HTMLElement) => {},
            })
          })
        },
        unobserve: (ele: HTMLElement) => {},
      }
    }

    renderWithModalRoot(
      <App>
        <SongMore
          albumName="abc"
          artistName="cdf"
          artistId={1}
          songName="ghj"
          albumId={2}
          songId={songId}
          imgUrl={"fake-url"}
        />
      </App>,
    )

    act(() => {
      fireEvent.click(screen.getByTestId("song-more"))
      jest.runOnlyPendingTimers()
    })
    await waitFor(() => {
      expect(screen.getByText("暂无版权")).toBeInTheDocument()
    })
    fireEvent.click(screen.getByText("暂无版权"))
  })

  it("can nav to artist page", async () => {
    renderWithModalRoot(
      <App>
        <SongMore
          albumName="abc"
          artistName="cdf"
          artistId={1}
          songName="ghj"
          albumId={2}
          songId={songId}
          imgUrl={"fake-url"}
        />
      </App>,
    )
    act(() => {
      fireEvent.click(screen.getByTestId("song-more"))
      jest.runAllTimers()
    })
    await waitFor(() => {
      expect(screen.getByText("下一首播放")).toBeInTheDocument()
    })

    act(() => {
      fireEvent.click(screen.getByText("歌手 cdf"))
      jest.runAllTimers()
    })
  })

  it("can nav to album page", async () => {
    // jest.useFakeTimers()
    renderWithModalRoot(
      <App>
        <SongMore
          albumName="abc"
          artistName="cdf"
          artistId={1}
          songName="ghj"
          albumId={2}
          songId={songId}
          imgUrl={"fake-url"}
        />
      </App>,
    )
    act(() => {
      fireEvent.click(screen.getByTestId("song-more"))
      jest.runOnlyPendingTimers()
    })

    await waitFor(() => {
      expect(screen.getByText("下一首播放")).toBeInTheDocument()
    })
    act(() => {
      fireEvent.click(screen.getByText("专辑 abc"))
      jest.runAllTimers()
    })
  })
})
