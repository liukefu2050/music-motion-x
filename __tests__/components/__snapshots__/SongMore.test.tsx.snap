// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongMore render well when in viewpoint 1`] = `
<DocumentFragment>
   
  <div
    class=" flex ml-auto relative after:absolute after:w-[40px] after:h-[30px] after:top-1/2 after:left-1/2 after:-translate-x-1/2 after:-translate-y-1/2"
    data-testid="song-more"
  >
    <span
      class=" w-1 h-1 rounded-[50%] bg-dg"
    />
    <span
      class=" w-1 h-1 rounded-[50%] bg-dg mx-1"
    />
    <span
      class=" w-1 h-1 rounded-[50%] bg-dg "
    />
  </div>
</DocumentFragment>
`;

exports[`SongMore render well when in viewpoint 2`] = `
<DocumentFragment>
   
  <div
    class=" flex ml-auto relative after:absolute after:w-[40px] after:h-[30px] after:top-1/2 after:left-1/2 after:-translate-x-1/2 after:-translate-y-1/2"
    data-testid="song-more"
  >
    <span
      class=" w-1 h-1 rounded-[50%] bg-dg"
    />
    <span
      class=" w-1 h-1 rounded-[50%] bg-dg mx-1"
    />
    <span
      class=" w-1 h-1 rounded-[50%] bg-dg "
    />
  </div>
  <div>
    <div
      class=" z-[9999999] fixed top-0 left-0 bottom-0 right-0 bg-mask bg-opacity-40"
      data-testid="modal-mask"
    >
      <div
        class=" bg-mg absolute bottom-0 pt-6 px-4 pb-36 transition-transform duration-200 w-full"
        style="border-radius: 12px 12px 0 0; transform: translate3d(0, 0,0);"
      >
        <div
          class="song_name text-fg text-base single_line"
        >
          ghj
        </div>
        <ul
          class="contents min-h-[200px] max-h-[55vh] overflow-y-auto"
        >
          <li
            class="next_song text-fg pl-11 mr-0 mb-0 ml-1 text-sm h-[30px] leading-[30px] mt-10"
            style="background-repeat: no-repeat; background-size: 23px 23px; background-image: url(test-file-stub);"
          >
            下一首播放
          </li>
          <li
            class="artist text-fg pl-11 mt-7 mr-0 mb-0 ml-1 text-sm h-[30px] leading-[30px] single_line"
            style="background-repeat: no-repeat; background-size: 23px 23px; background-image: url(test-file-stub);"
          >
            歌手 cdf
          </li>
          <li
            class="album text-fg pl-11 mt-7 mr-0 mb-0 ml-1 text-sm h-[30px] leading-[30px] single_line"
            style="background-repeat: no-repeat; background-size: 23px 23px; background-image: url(test-file-stub);"
          >
            专辑 abc
          </li>
        </ul>
        <div
          class="close absolute bottom-10 text-fg text-center text-3xl w-full left-0"
          data-close="true"
        >
          ×
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
